url8f84d8d956_klapaukh_JTuttePoly_tgz-pJ8-BitTutteJ8/src/graph/Util.java

```diff
@@ -0,0 +1,210 @@
+package graph;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+
+import util.Pair;
+
+// NOTES: only works for connected graphs at the moment
+
+public final class Util {
+	
+	public static boolean isomorphic(Graph g1, Graph g2) {
+		// First, apply simple tests
+		if (g1.numVertices() != g2.numVertices() || g1.numEdges() != g2.numEdges()) {
+			return false;
+		}			
+		
+		// Second, select spine for g2 and partition
+		ArrayList<BitSet> g2_partitions = null;
+		int g2_partitions_size = Integer.MIN_VALUE;
+		int w=0;				
+//		for(int i=0;i!=g2.size();++i) {
+		for(int i: g2.vertices()) {
+			ArrayList<BitSet> partitions = initialPartition(i,g2);
+			if(partitions.size() > g2_partitions_size) {
+				w = i;
+				g2_partitions = partitions;
+				g2_partitions_size = g2_partitions.size();
+			}
+		}
+		
+		int w_degree = g2.numEdges(w);
+		partition(w,g2_partitions,g2);		
+		Matrix g2_normalised = normalise(g2_partitions,g2);
+		
+		// Third, try to find candidates to check off
+		
+		outer:
+//		for(int i=0;i!=g1.size();++i) {
+		for(int i: g1.vertices()) {
+			ArrayList<BitSet> g1_partitions = initialPartition(i,g1);
+			int g1_partitions_size = g1_partitions.size(); 
+			
+			if(g1_partitions_size > g2_partitions_size) {
+				// definitely not isomorphic
+				break;
+			} else if (g1_partitions_size == g2_partitions_size
+					&& g1.numEdges(i) == w_degree) {
+				
+				// partition the candidate
+				partition(i,g1_partitions,g1);
+				
+				// check partition cardinality
+				for(int j=0;j!=g1_partitions.size();++j) {
+					BitSet g1_partition = g1_partitions.get(j);
+					BitSet g2_partition = g2_partitions.get(j);
+					if(g1_partition.cardinality() != g2_partition.cardinality()) {
+						// hmmm, yuk
+						continue outer;
+					}
+				}
+				
+				// normalise and check candidate
+				Matrix g1_normalised = normalise(g1_partitions,g1);				
+				
+				if(g1_normalised.equals(g2_normalised)) {
+					return true;
+				}
+			}
+		}
+		
+		return false;
+	}
+	
+	public static Matrix normalise(ArrayList<BitSet> partitions, Graph g) {
+		int[] rpartitions = reversePartitions(partitions,g);
+		Matrix weights = new Matrix(partitions.size(),partitions.size());
+		
+		for(int i=0;i!=partitions.size();++i) {
+			BitSet partition = partitions.get(i);
+			int rep = partition.nextSetBit(0);			
+			
+			for(Pair<Integer,Integer> w : g.edges(rep)) {
+				int pIndex = rpartitions[w.first()];
+				int weight = weights.get(i,pIndex);
+				weights.set(weight+1,i,pIndex);
+			}			
+		}
+		
+		return weights;
+	}
+	
+	public static void partition(int v, ArrayList<BitSet> partitions, Graph g) {
+		boolean changed = true;
+		while(changed) {
+			changed = false;
+			for(int i=0;i!=partitions.size();++i) {
+				if(split(i,partitions,g)) {
+					changed = true;
+					break;
+				}
+			}
+		}		
+	}
+	
+	public static boolean split(int index, ArrayList<BitSet> partitions, Graph g) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.